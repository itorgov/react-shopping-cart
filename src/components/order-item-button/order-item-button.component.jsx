import React from 'react';
import classNames from 'classnames';

import {ReactComponent as IconTrash} from './trash.svg';
import {ReactComponent as IconPlus} from "./plus.svg";
import {ReactComponent as IconMinus} from "./minus.svg";
import styles from './order-item-button.module.scss';

const renderIcon = variant => {
    switch (variant) {
        case 'remove':
            return <IconTrash/>;
        case 'increment':
            return <IconPlus/>;
        case 'decrement':
            return <IconMinus/>;
        default:
            return '';
    }
};

export const OrderItemButton = ({variant, shake = false}) => (
    <button className={classNames({
        [styles.button]: true,
        [styles[variant]]: true,
        [styles.shake]: shake,
    })}
    >
        {renderIcon(variant)}
    </button>
);
