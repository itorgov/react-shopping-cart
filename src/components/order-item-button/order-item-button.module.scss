.button {
  background-color: transparent;
  border: 0;

  display: flex;
  justify-content: center;
  align-items: center;

  &:focus {
    outline: none;
  }

  & svg {
    width: 2rem;
    height: 2rem;
    fill: currentColor;
    opacity: 0.8;
    transition: all 0.2s;
  }

  &:hover svg,
  &:active svg,
  &:focus svg {
    opacity: 1;
  }
}

.remove {
  color: var(--color-secondary);
}

.increment,
.decrement {
  color: var(--color-text-dark);

  & svg {
    opacity: 0.6;
  }
}

.shake {
  &:hover,
  &:focus {
    animation: shake 1s;
  }
}

@keyframes shake {
  0% {
    transform: rotate(0);
  }

  25% {
    transform: rotate(10deg);
  }

  50% {
    transform: rotate(-10deg);
  }

  75% {
    transform: rotate(5deg);
  }

  100% {
    transform: rotate(0);
  }
}
