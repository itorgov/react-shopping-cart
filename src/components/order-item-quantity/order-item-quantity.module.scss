.quantity {
  margin: 0 1rem;
}
