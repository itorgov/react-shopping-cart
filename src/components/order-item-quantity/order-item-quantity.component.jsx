import React from 'react';
import {useDispatch} from "react-redux";

import {OrderItemButton} from "../order-item-button/order-item-button.component";

import styles from './order-item-quantity.module.scss';

export const OrderItemQuantity = ({item}) => {
    const dispatch = useDispatch();

    return (
        <div>
            <OrderItemButton
                variant='decrement'
                onClick={() => dispatch({type: 'DECREMENT_ITEM_QUANTITY', payload: item.id})}
            />
            <span className={styles.quantity}>{item.quantity}</span>
            <OrderItemButton
                variant='increment'
                onClick={() => dispatch({type: 'INCREMENT_ITEM_QUANTITY', payload: item.id})}
            />
        </div>
    );
};
