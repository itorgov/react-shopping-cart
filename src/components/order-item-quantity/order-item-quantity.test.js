import React from 'react'
import {mount} from 'enzyme';
import {Provider} from 'react-redux';
import configureStore from 'redux-mock-store'

import {OrderItemQuantity} from './order-item-quantity.component'
import {OrderItemButton} from '../order-item-button/order-item-button.component'

let wrapper;
let store;

beforeEach(() => {
    const mockStore = configureStore();
    const mockItem = {
        id: 1,
        name: "Test name",
        price: 3900,
        quantity: 2,
    };
    const initialState = {
        items: [mockItem],
        note: '',
    };
    store = mockStore(initialState);
    wrapper = mount(<Provider store={store}><OrderItemQuantity item={mockItem} /></Provider>);
});

describe('<OrderItemQuantity />', () => {
    it('expect to render', () => {
        expect(wrapper.find('span').text()).toEqual('2');
        expect(wrapper.find(OrderItemButton).length).toEqual(2);
        expect(wrapper).toMatchSnapshot();
    });

    it('decrementing button has to dispatch DECREMENT_ITEM_QUANTITY action', () => {
        expect(wrapper.find(OrderItemButton).at(0).prop('variant')).toBe('decrement');

        wrapper.find(OrderItemButton).at(0).simulate('click');

        const action = store.getActions();
        expect(action[0].type).toBe("DECREMENT_ITEM_QUANTITY");
    });

    it('incrementing button has to dispatch INCREMENT_ITEM_QUANTITY action', () => {
        expect(wrapper.find(OrderItemButton).at(1).prop('variant')).toBe('increment');

        wrapper.find(OrderItemButton).at(1).simulate('click');

        const action = store.getActions();
        expect(action[0].type).toBe("INCREMENT_ITEM_QUANTITY");
    });
});
