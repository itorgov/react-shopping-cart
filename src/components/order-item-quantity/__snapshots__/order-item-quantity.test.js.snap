// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<OrderItemQuantity /> expect to render 1`] = `
<Provider
  store={
    Object {
      "clearActions": [Function],
      "dispatch": [Function],
      "getActions": [Function],
      "getState": [Function],
      "replaceReducer": [Function],
      "subscribe": [Function],
    }
  }
>
  <OrderItemQuantity
    item={
      Object {
        "id": 1,
        "name": "Test name",
        "price": 3900,
        "quantity": 2,
      }
    }
  >
    <div>
      <OrderItemButton
        onClick={[Function]}
        variant="decrement"
      >
        <button
          className="button decrement"
          onClick={[Function]}
        >
          <ForwardRef(SvgMinus)>
            <svg>
              minus.svg
            </svg>
          </ForwardRef(SvgMinus)>
        </button>
      </OrderItemButton>
      <span
        className="quantity"
      >
        2
      </span>
      <OrderItemButton
        onClick={[Function]}
        variant="increment"
      >
        <button
          className="button increment"
          onClick={[Function]}
        >
          <ForwardRef(SvgPlus)>
            <svg>
              plus.svg
            </svg>
          </ForwardRef(SvgPlus)>
        </button>
      </OrderItemButton>
    </div>
  </OrderItemQuantity>
</Provider>
`;
